protocol/msgpack/base_iterator.go

```diff
@@ -251,6 +270,69 @@ func (it *baseIterator) decodeFloat64() float64 {
 	return value
 }
 
+func (it *baseIterator) decodeFloat64Slice(encodingType encodingType) ([]float64, pool.FloatsPool) {
+	if encodingType == nonPackedEncoding {
+		return it.decodeFloat64SliceNative()
+	}
+	return it.decodeFloat64SlicePacked()
+}
+
+func (it *baseIterator) decodeFloat64SliceNative() ([]float64, pool.FloatsPool) {
+	numValues := it.decodeArrayLen()
+	if it.decodeErr != nil {
+		return nil, nil
+	}
+	decoded, pool := it.getFloat64SliceFor(numValues)
+	for i := 0; i < numValues; i++ {
+		decoded = append(decoded, it.decodeFloat64())
+	}
+	if it.decodeErr != nil && pool != nil {
+		pool.Put(decoded)
+		return nil, nil
+	}
+	return decoded, pool
+}
+
+func (it *baseIterator) decodeFloat64SlicePacked() ([]float64, pool.FloatsPool) {
+	numBytes := it.decodeBytesLen()
+	if it.decodeErr != nil {
+		return nil, nil
+	}
+	numValues := numBytes / numBytesInFloat64
+	decoded, pool := it.getFloat64SliceFor(numValues)
+	for i := 0; i < numValues; i++ {
+		_, err := it.reader().Read(it.tmpBuf)
+		if err != nil {
+			it.decodeErr = err
+			if pool != nil {
+				pool.Put(decoded)
+			}
+			return nil, nil
+		}
+		decoded = append(decoded, math.Float64frombits(byteOrder.Uint64(it.tmpBuf)))
+	}
+	return decoded, pool
+}
+
+// getFloat64SliceFor prepares a zero-length slice with sufficient capacity
+// for numValues floating point numbers.
+func (it *baseIterator) getFloat64SliceFor(numValues int) ([]float64, pool.FloatsPool) {
+	if cap(it.cachedFloat64s) >= numValues {
+		it.cachedFloat64s = it.cachedFloat64s[:0]
+		return it.cachedFloat64s, nil
+	}
+	if numValues <= it.largeFloatsSize {
+		newCapcity := int(math.Max(float64(numValues), float64(cap(it.cachedFloat64s)*2)))
+		if newCapcity > it.largeFloatsSize {
```

Comment:
Why not just the capacity of `it.cachedFloat64s` to `it.largeFloatSize` when we initialize the iterator? It seems like after enough time it will reach that value, no?